contrib/pkg/sets/v2/sets.go

```diff
@@ -61,200 +85,249 @@ func (r *ResourceDelta[T]) DeltaV1() sk_sets.ResourceDelta {
+		keys.Insert(sk_sets.Key(resource))
+	}
+	return sets.Set[string]{}
 }
 
-func (s *resourceSet[T]) List(filterResource ...func(T) bool) []T {
+// Filter returns an iterator that will iterate over the set of elements
+// that match the provided filter. If the filter returns true, the resource will be included in the iteration.
+func (s *resourceSet[T]) Filter(filterResource func(T) bool) func(yield func(int, T) bool) {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
-	var resources []T
-	for _, key := range s.set.List() {
-		var filtered bool
-		for _, filter := range filterResource {
-			if filter(s.mapping[key]) {
-				filtered = true
-				break
+	return func(yield func(int, T) bool) {
+		i := 0
+		for _, resource := range s.set {
+			if filterResource(resource) {
+				if !yield(i, resource) {
+					break
+				}
+				i += 1
 			}
 		}
-		if !filtered {
-			resources = append(resources, s.mapping[key])
-		}
 	}
-	return resources
 }
 
-func (s *resourceSet[T]) UnsortedList(filterResource ...func(T) bool) []T {
+// This is a convenience function that constructs a list of resource that do not match any of the provided filters.
+// Use of this function should be limited to only when a filtered list is needed, as this allocates a new list.
+func (s *resourceSet[T]) FilterOutAndCreateList(filterResource ...func(T) bool) []T {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
-
-	keys := s.set.UnsortedList()
-	resources := make([]T, 0, len(keys))
-
-	for _, key := range keys {
-		var filtered bool
+	var ret []T
+	for _, resource := range s.set {
+		filtered := false
 		for _, filter := range filterResource {
-			if filter(s.mapping[key]) {
+			if filter(resource) {
 				filtered = true
 				break
 			}
 		}
 		if !filtered {
-			resources = append(resources, s.mapping[key])
+			ret = append(ret, resource)
+		}
+	}
+	return ret
+}
+
+// Iter iterates over the set, passing the index and resource to the provided function for every element in the set.
+// The iteration can be stopped by returning false from the function. This can be thought of as a "break" statement in a loop.
+// Returning true will continue the iteration. This can be thought of as a "continue" statement in a loop.
+func (s *resourceSet[T]) Iter(yield func(int, T) bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	for i, resource := range s.set {
+		if !yield(i, resource) {
+			break
 		}
 	}
-	return resources
 }
 
 func (s *resourceSet[T]) Map() map[string]T {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
 	newMap := map[string]T{}
-	for k, v := range s.mapping {
-		newMap[k] = v
+	for _, resource := range s.set {
+		newMap[sk_sets.Key(resource)] = resource
 	}
 	return newMap
 }
 
-func (s *resourceSet[T]) Insert(
-	resources ...T,
-) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+// Insert adds items to the set in inserted order.
+// If an item is already in the set, it is overwritten.
+// The set is sorted based on the ResourceId of the resources.
+func (s *resourceSet[T]) Insert(resources ...T) {
```

Comment:
we're changing insertion and lookup to be `O(n·log(n))`instead of `O(1)`. Did we verify this doesn't cause perf issues?